spell_ids.tpp

```diff
@@ -14,6 +15,3144 @@ FOR (i = 1000; i < 5000; ++i) BEGIN
 	PATCH_IF (FILE_EXISTS_IN_GAME ~%spell_res%.spl~) BEGIN
 		PATCH_IF !(IS_AN_INT ~%spell_name%~) BEGIN	// For instance, SPCL643 exists in BG:EE but does not have a symbolic reference...
 			TEXT_SPRINT EVAL ~%spell_name%~ ~%spell_res%~
+			PATCH_LOG "TEXT_SPRINT %spell_name% ~%spell_res%~"
 		END
 	END
 END
+*/
+
+PATCH_IF (GAME_IS "BGEE BG2EE") BEGIN
```

Comment:
Why this restriction? What about classic?

Reply:
> Why this restriction? What about classic?

Because SPELL.IDS have different entries depending on the game: for instance, IWD:EE has `CLERIC_SUNSCORCH` whereas BG(2):EE doesn't...

As far as the classic games are concerned: I do not have them, so I cannot check their `SPELL.IDS` file. I mean, I expect it to be more or less the same as EE games, but you might want to run `PATCH_LOG` yourself and see how much they differ...

Having said that, how should we treat these things that differ from game to game?

Reply:
> As far as the classic games are concerned: I do not have them, so I cannot check their `SPELL.IDS` file. I mean, I expect it to be more or less the same as EE games, but you might want to run `PATCH_LOG` yourself and see how much they differ...

They are on [IESDP](https://iesdp.bgforge.net/files/ids/bg2/index.htm).

> Having said that, how should we treat these things that differ from game to game?

- There's no problem with having CLERIC_SUNSCORCH in completion for a BG2 mod. I mean, you could use it in code and maybe get a bug, but that's hardly the tool's fault. There's no ambiguity or misleading.
- There's also no problem with having two different symbols from different IDS resolve to the same value. It will evaluate and compile just fine.
- However, having two different values from two different IDS for the same symbol would be an issue. I hope that's not the case?

Reply:
Yes, of course auto-completion must have **all** those entries (irrespective of the game in question). I added a `GAME_IS` check in order not to have `CLERIC_SUNSCORCH` (and the like) set to a non-existing resource for a non-IWDEE mod... But if you're saying this is not that bad, then must I remove all those `GAME_IS` checks? If so, then I should also remove all those entries that are in common or they'll get set twice (which is pretty stupid...)

> There's also no problem with having two different symbols from different IDS resolve to the same value. It will evaluate and compile just fine.

This happens with opcode #318... Speaking of this, should I remove the `GAME_IS` check in this case?

> However, having two different values from two different IDS (games) for the same symbol would be an issue. I hope that's not the case?

The IWDification mod introduces this issue with the IWD version of `WIZARD_CONJURE_FIRE_ELEMENTAL` (`EARTH` and `AIR` too). It basically renames 
`WIZARD_CONJURE_LESSER_*_ELEMENTAL` as `WIZARD_CONJURE_*_ELEMENTAL` (probably because in IWD games there is no level 6 version of this spell). But that's not good because in BG games there **is** the level 6 version, so now you have two different `SPL` files that point to the same identifier! I mean, since you're replacing the level 5 series, then you should leave the symbolic references as they are.

Having said that, all of this is probably irrelevant for this library because since the list is static, the library will correctly assign `WIZARD_CONJURE_FIRE_ELEMENTAL` to `SPWI620` instead of `SPWI516`...

Reply:
I think all is good then, and game version checks can be just dropped.
Maybe somewhere down the line MLS will support engine dialects, but that'll require more changes than just `tpp`, and for now worrying about it is quire premature.

Reply:
@burner1024 

Ah, wait: `WIZARD_CONJURE_FIRE_ELEMENTAL` is `SPWI620` in BG games; in IWDEE is `SPWI531`...

Reply:
Well, then I guess we can wrap just the differing values with checks, to avoid duplication.

Reply:
> Well, then I guess we can wrap just the differing values with checks, to avoid duplication.

OK, it's gonna take a while though... You know, it's a pretty boring task...

Reply:
Why, it's just one spell, or there are more?

Reply:
> Why, it's just one spell, or there are more?

There are more.

For instance, `INNATE_FLASH_DESTROY_SUMMON_YXUNOMEI`, `INNATE_FLASH_DESTROY_SUMMON_BELHIFET`, `INNATE_MYCONID_SPORES` etc. are present only in IWDEE...

Reply:
Let me check.

Reply:
Yes, the IDS are a mess and it's chore to sort them all out. I suggest to keep them as is (complete), just in different files, and then do something like `PATCH_IF (GAME_IS "BGEE BG2EE") BEGIN INCLUDE spell_ids_bgee.tpp END` and so on.

There will be a small issue with hover/completion, as it will use the last value for tooltip. But that's purely visual, and I think good enough for now.

Reply:
> There will be a small issue with hover/completion, as it will use the last value for tooltip. But that's purely visual, and I think good enough for now.

What about adding all variants?
```
BG games: SPWI620
IWDEE: SPWI531

IElib define
```
Is that possible? If so, it requires some sort of sorting though...

Reply:
Possible, yes. But assuming that we want to display variants only for constants that differ, that would be somewhat hacky and maybe not worth the trouble. (Because the files are parsed alphabetically and we don't know in advance whether there's a dupe).
Here's the [script](https://github.com/BGforgeNet/VScode-BGforge-MLS/blob/master/scripts/ie-update.py) that does the parsing, and [here's](https://github.com/BGforgeNet/VScode-BGforge-MLS/blob/1df9886a5f9ff1c28793896e53afe280dd8148bf/server/out/weidu.completion.yml#L1892) where completion is saved.
Thing is, BGforge mods are pretty much exclusively BG, so this is more of a theoretical issue for me. And I prefer to focus on more pressing ones.

All of this would be easier, more robust and streamlined if IESDP was machine readable, like we discussed in https://github.com/BGforgeNet/VScode-BGforge-MLS/pull/18...

Reply:
> ...and we don't know in advance whether there's a dupe...

What do you know exactly?

We **do** know in advance that `WIZARD_CONJURE_FIRE_ELEMENTAL` points to `SPWI620` in BG games and to `SPWI531` in IWD...

Reply:
The linked script parses the files and finds values sequentially.

Reply:
> The linked script parses the files and finds values sequentially.

So, writing something like this would be totally pointless?
```
- name: WIZARD_CONJURE_FIRE_ELEMENTAL
   detail: 'BG games: SPWI620; IWDEE: SPWI531'
   doc: IElib define
```

Reply:
> So, writing something like this would be totally pointless?

Sorry, I don't quite understand the question. i guess it could be useful to someone. Or could be not.

Reply:
> > So, writing something like this would be totally pointless?
> 
> Sorry, I don't quite understand the question. i guess it could be useful to someone. Or could be not.

Sorry, the problem is that I didn't understand why this fact is relevant:
_"The linked script parses the files and finds values sequentially."_

Reply:
Because latter symbols will override the former ones.
Anyway, I think we're too hang up on this minor issue. Let's just go with however it'll work for now, and revisit if/when someone actually comes with a live problem.